source/simulator/helper_functions.cc

```diff
@@ -1484,6 +1484,212 @@ namespace aspect
+
+              heating_model_manager.evaluate(in_C, out_C, heating_model_outputs_C);
+
+              for (unsigned int j=0; j<dof_handler.get_fe().base_element(introspection.base_elements.compositional_fields).dofs_per_cell; ++j)
+                {
+                  for (unsigned int c=0; c<introspection.n_compositional_fields; ++c)
+                    {
+                      // simple forward euler
+                      in_C.composition[j][c] = in_C.composition[j][c]
+                                               + reaction_time_step_size * reaction_rate_outputs_C->reaction_rates[j][c];
+                      accumulated_reactions_C[j][c] += reaction_time_step_size * reaction_rate_outputs_C->reaction_rates[j][c];
+                    }
+                  in_C.temperature[j] = in_C.temperature[j]
+                                        + reaction_time_step_size * heating_model_outputs_C.rates_of_temperature_change[j];
+                }
+
+              // loop over temperature element
+              material_model->evaluate(in_T, out_T);
+
+              heating_model_manager.evaluate(in_T, out_T, heating_model_outputs_T);
+
+              for (unsigned int j=0; j<dof_handler.get_fe().base_element(introspection.base_elements.temperature).dofs_per_cell; ++j)
+                {
+                  // simple forward euler
+                  in_T.temperature[j] = in_T.temperature[j]
+                                        + reaction_time_step_size * heating_model_outputs_T.rates_of_temperature_change[j];
+                  accumulated_reactions_T[j] += reaction_time_step_size * heating_model_outputs_T.rates_of_temperature_change[j];
+
+                  for (unsigned int c=0; c<introspection.n_compositional_fields; ++c)
+                    in_T.composition[j][c] = in_T.composition[j][c]
+                                             + reaction_time_step_size * reaction_rate_outputs_T->reaction_rates[j][c];
+                }
+            }
+
+          // copy reaction rates and new values for the compositional fields
+          for (unsigned int j=0; j<dof_handler.get_fe().base_element(introspection.base_elements.compositional_fields).dofs_per_cell; ++j)
+            for (unsigned int c=0; c<introspection.n_compositional_fields; ++c)
+              {
+                const unsigned int composition_idx
+                  = dof_handler.get_fe().component_to_system_index(introspection.component_indices.compositional_fields[c],
+                                                                   /*dof index within component=*/ j);
+
+                // skip entries that are not locally owned:
+                if (!dof_handler.locally_owned_dofs().is_element(local_dof_indices[composition_idx]))
+                  continue;
+
+                distributed_vector(local_dof_indices[composition_idx]) = in_C.composition[j][c];
+                distributed_reaction_vector(local_dof_indices[composition_idx]) = accumulated_reactions_C[j][c];
```

Comment:
Oh, I meant to post this below your comment above.

> whether we are setting the same DoFs twice

Correct, this will happen. I think the way we originally designed the algorithm in Georgia we accounted for that (correct result but computing the ODEs more often than necessary). I am not sure how it is working now.